### ISSUE TITLE
[Feature Request] Docs showing how to use this with NestJS & its GraphQL module

### ISSUE THREAD

**Issue by jamesopti** [davidje13/superwstest#6]
I'm trying to get this up and running for GraphQL subscriptions running via my NestJS Apollo Graphql server.

No matter what I do, the websocket connection is immediately in CLOSING state.

Any guidance or docs on this would be appreciated!

Here's what I've got so far:

```ts
import requestWS from 'superwstest'
import * as WebSocket from 'ws'

...

describe('(e2e)', () => {
  let app: INestApplication
  let client: PrismaClient

  beforeAll(async () => {
    const moduleFixture: TestingModule = await Test.createTestingModule({
      imports: [
        ConfigModule,
        GraphQLModule.forRootAsync({
          imports: [ConfigModule],
          inject: [ConfigService],
          useFactory: async (configService: ConfigService) => ({
            typePaths: ['./**/*.graphql'],
            introspection: true,
            context: ({ req, connection }) => {
              return connection ? { req: connection.context } : { req }
            },
            subscriptions: {
              onConnect: getOnConnect(configService),
            },
            installSubscriptionHandlers: true,
          }),
        }),
        IdentityModule,
      ],
      providers: [PrismaService],
    }).compile()

    prismaClient = new PrismaClient()

    app = moduleFixture.createNestApplication()
    app.use(cookieSession())
    app.use(initialize())
    app.use(session())
    await app.init()
    const configService = app.get<ConfigService>(ConfigService)
    const port = configService.get<number>('app.port')
    await new Promise((res) => {
      app.listen(port, () => {
        console.log('SERVER LISTENING')
        res(null)
      })
    })
  })

  afterAll(async () => {
    await prismaClient.$disconnect()
    await app.close()
  })

  afterEach(async () => {
  })

  describe('Subscriptions', () => {
    it('should succeed in connecting', async () => {
      const cookie = await fakeSessionCookie(user)
      const ws = await requestWS(app.getHttpServer(), { shutdownDelay: 500 })
        .ws('/graphql', { headers: { cookie } })
        .expectUpgrade((res) => {
          console.log('UPGRADE', res)
          return true
        })

      expect(ws.readyState).toEqual(WebSocket.OPEN)
    })
  })
})
```

**Comment by davidje13** [comment]
I haven't used NestJS much, and never with websockets, but is it possible that `app.getHttpServer()` isn't listening for websocket connections? (maybe the test harness doesn't support websockets, or exposes them in a different server?) Do you see your `UPGRADE` log line printed?

Your actual test code looks fine to me, but obviously there are a number of unknowns here. For example maybe the `fakeSessionCookie` isn't in the correct format? Could you try the same with authentication disabled to rule it out?

Also you may find it useful to check the close reason. You could do something like this to see it:

```js
const ws = await requestWS(app.getHttpServer())
  .ws('/graphql', { headers: { cookie } })
  .expectClosed(0, ''); // will fail here, but error message should tell you which code and reason was sent by the server
```

(`expectClosed` will wait until the connection is closed by the server and optionally check the code and message. Since that won't be `0`, this code will cause it to print what it actually got as part of the error message. If it is _not_ closed by the server, Jest will eventually time out the test)

**Comment by jamesopti** [comment]
Thanks for the quick reply @davidje13 ! - Lemme answer inline:

> Do you see your `UPGRADE` log line printed?

Yes! When I console log the entire `ws` instance though, I can tell that it's closed already.

> Could you try the same with authentication disabled to rule it out?

Yup, getting the same result with no auth.

> Also you may find it useful to check the close reason. You could do something like this to see it:

`Expected close code 0, got 1002` - Im not familiar with websocket error codes but it seems like this is a protocol error? I'm not sending any data though which is interesting.

**Comment by jamesopti** [comment]
> I haven't used NestJS much, and never with websockets, but is it possible that `app.getHttpServer()` isn't listening for websocket connections?

I'm suspicious of this too and am now looking into that based on this nest issue: https://github.com/nestjs/nest/issues/3463

**Comment by davidje13** [comment]
looks like the issue you found was just because the person was mixing websockets with socketIO (socketIO can use websockets, among other connection options, but builds extra protocols on top, so they're not directly compatible).

Code 1002 is supposed to be for low-level protocol errors but I can't see why that would apply here. In practice servers are able to use any code for any reason, so I suspect something in the NestJS/GraphQL chain is deciding to use it for an application-related error instead. It probably means the server expects you to connect with a specific subprotocol. By default, no subprotocols are used, but you can specify them on the `.ws` call:

```javascript
const ws = await requestWS(app.getHttpServer())
  .ws('/graphql', ['some-sub-protocol-here', 'more-than-one-if-you-need-to'], { headers: { cookie } })
```

(subprotocols are arbitrary strings which the server can recognise and change behaviour accordingly)

You'd need to check how you connect to it from a non-test client and match that (or if you're using a library to do that, you'll need to check the docs to see if it needs a particular subprotocol, as well as the details of what messages look like in that protocol).

---

I pushed a new version (1.4.0) which includes the close code and reason in all failure messages. It might help you debug here to see the message (1002 is the code but the server might also be sending a human-readable message which explains more about what's going on)

**Comment by jamesopti** [comment]
> ```js
> const ws = await requestWS(app.getHttpServer())
>   .ws('/graphql', ['some-sub-protocol-here', 'more-than-one-if-you-need-to'], { headers: { cookie } })
> ```
> 
> (subprotocols are arbitrary strings which the server can recognise and change behaviour accordingly)

THANK YOU so much! Your hunch was dead on. Some googling led me to the underlying websocket library being used, and sure enough they use their own `graphql-ws` subprotocol, which has me off and running!

https://github.com/apollographql/subscriptions-transport-ws/blob/c6fb351fe3d7c7a447438798fab43c268b33036e/src/protocol.ts


I'll follow up here once I have my test more fleshed out, but thanks again so much for the quick replies and even publishing a new version!

**Comment by davidje13** [comment]
Glad I managed to help :)

**Comment by davidje13** [comment]
Closing this since the issue is resolved. It won't be locked so you can still add more comments later if you want.

**Comment by guiaramos** [comment]
@jamesopti I am facing the same issue, could you please help me out with your solution?

```ts
 const ws = await requestWS(app.getHttpServer())
            .ws(GRAPHQL_WS_PATH, ['graphql-ws'])
            .expectUpgrade((res) => {
              console.log('UPGRADE', res);
              return true;
            });
```

and getting the following error:

```bash
Server must be listening:
    beforeEach((done) => server.listen(0, done));
    afterEach((done) => server.close(done));

    supertest's request(app) syntax is not supported (find out more: https://github.com/davidje13/superwstest#why-isnt-requestapp-supported)
```

my graphql settings: 
```ts
{
      autoSchemaFile: join(process.cwd(), 'schema.gql'),
      debug: false,
      installSubscriptionHandlers: true,
      path: GRAPHQL_PATH,
      cors: corsOptions,
      context: ({ req, res, payload, connection }) => ({
        req,
        res,
        payload,
        connection,
      }),
      subscriptions: {
        'graphql-ws': {
          path: GRAPHQL_WS_PATH,
        },
      },
      playground: false,
      plugins: [ApolloServerPluginLandingPageLocalDefault()],
    };
```

**Comment by guiaramos** [comment]
Also, I am using Fastify

**Comment by guiaramos** [comment]
@jamesopti thanks for you reply!

I was able to fix it by calling the method listen inside of beforeEach
```ts
describe('VerificationTokenController (e2e)', () => {
  let moduleFixture: TestingModule;
  let app: INestApplication;
  let config: ConfigService;

  beforeAll(async () => {
    moduleFixture = await Test.createTestingModule({
      imports: [AppModule],
    }).compile();

    app = moduleFixture.createNestApplication(new FastifyAdapter());
    app.useGlobalPipes(
      new ValidationPipe({
        exceptionFactory,
        forbidUnknownValues: false,
      }),
    );

    await app.init();
    await app.getHttpAdapter().getInstance().ready();
    await app.listen(4000, '0.0.0.0');

    config = app.get(ConfigService);
  });
....
})
```

**Comment by davidje13** [comment]
Sorry for the late reply here. The issue you faced (@guiaramos & @jamesopti) is more closely connected to #11 than this one, and is due to superwstest not supporting auto-starting the server (you must explicitly start the server). For the reasons behind this you can see the issue I linked to, or the relevant part of the documentation (linked in the error message itself).

Don't forget to also _close_ the server when you're done (in an `afterAll` or `afterEach`) or there's a risk your tests will hang at the end due to servers still running. Also I'd suggest not listening on `0.0.0.0` since that will make your server accessible on the local network, which could be a risk depending on your app behaviour and firewall settings (use `localhost`, `127.0.0.1`, or `::1` instead. If you have problems with `localhost` or `::1` you can upgrade to version 1.9.0 which fixes an IPv6 issue)

Do you have any suggestions on how to make the error message clearer? I thought it was fairly unambiguous but then I've already got a good understanding of how this works internally, so it's always good to hear an outside perspective.

**Comment by guiaramos** [comment]
@davidje13 thank you for your reply and for the advices!

**Comment by Shinigami92** [comment]
It took me around 3h to finally get it to work

Here is an example of my e2e test:

```ts
import type { INestApplication } from '@nestjs/common';
import { ValidationPipe } from '@nestjs/common';
import type { TestingModule } from '@nestjs/testing';
import { Test } from '@nestjs/testing';
import { AppModule } from 'src/app.module';
import { CatService } from 'src/cats/cats.service';
import request from 'superwstest';

describe('CatsResolver (e2e)', () => {
  let app: INestApplication;

  beforeEach(async () => {
    const moduleFixture: TestingModule = await Test.createTestingModule({
      imports: [AppModule],
    }).compile();

    app = moduleFixture.createNestApplication();
    app.useGlobalPipes(new ValidationPipe());
    await app.init();

    await app.listen(3020, '0.0.0.0');
  });

  afterEach(async () => {
    await app.close();
  });

  // ...

  it('updateCat', async () => {
    const cats = await app
      .get(CatsService)
      .findAll({ limit: 1, offset: 0 });

    const id = cats[0].id;
    const name = 'Updated Cat';

    const subscriptionTest = request(app.getHttpServer())
      .ws('/graphql', ['graphql-ws'])
      .sendJson({ type: 'connection_init', payload: {} }) // Send the init message
      .sendJson({
        id: '1',
        type: 'start',
        payload: {
          operationName: 'catUpdated',
          query: `#graphql
          subscription catUpdated {
            catUpdated {
              id
              name
            }
          }`,
        },
      }) // Subscribe to the PubSub event
      .expectJson({ type: 'connection_ack' }); // This is the answer that the connection is working

    await subscriptionTest; // This is super important!!!

    // Make your request
    await request(app.getHttpServer())
      .post('/graphql')
      .send({
        operationName: 'updateCat',
        query: `#graphql
        mutation updateCat($id: String!, $data: UpdateCatInput!) {
          updateCat(id: $id, updateCatData: $data) {
            id
            name
          }
        }`,
        variables: {
          id,
          data: {
            name,
          },
        },
      })
      .expect(200)
      .expect(({ body }) => {
        const data = body.data;
        expect(data.updateCat).toBeTruthy();
        expect(data.updateCat).toEqual({
          id,
          name,
        });
      });

    await subscriptionTest
      .expectJson((message) => {
        expect(message.type).toEqual('data');
        const data = message.payload.data;
        expect(data).toEqual({
          catUpdated: {
            id,
            name,
          },
        });
      }) // Check that the PubSub event was fired
      .close()
      .expectClosed();
  });

  // ...
});
```

Following **relevant** deps were used (so I removed some not relevant):

```
  "dependencies": {
    "@nestjs/common": "~8.2.6",
    "@nestjs/core": "~8.2.6",
    "@nestjs/graphql": "~9.1.2",
    "@nestjs/platform-express": "~8.2.6",
    "apollo-server-express": "~3.6.2",
    "class-transformer": "~0.5.1",
    "class-validator": "~0.13.2",
    "graphql": "~15.8.0",
    "graphql-subscriptions": "~2.0.0",
    "reflect-metadata": "~0.1.13",
    "rxjs": "~7.5.2"
  },
  "devDependencies": {
    "@apollo/gateway": "0.46.0",
    "@nestjs/cli": "~8.2.0",
    "@nestjs/schematics": "~8.0.5",
    "@nestjs/testing": "~8.2.6",
    "@types/express": "~4.17.13",
    "@types/jest": "~27.4.0",
    "@types/node": "~16.11.22",
    "@types/supertest": "~2.0.11",
    "@types/ws": "~8.2.2",
    "jest": "~27.5.1",
    "jest-junit": "~13.0.0",
    "supertest": "~6.2.2",
    "superwstest": "~2.0.0",
    "typescript": "~4.5.5"
  }
```

I hope this will help someone out there and reduce they time <3

**Comment by davidje13** [comment]
@Shinigami92 I'm glad you got it working, but I'll just note that superwstest isn't designed to have multiple `await`s on a single chain, and I can't guarantee that will work in all future versions. What you're really looking for is `exec` ([docs](https://github.com/davidje13/superwstest#execfn)):

```javascript
it('updateCat', async () => {
  // ...

  await request(app.getHttpServer())
    .ws('/graphql', ['graphql-ws'])
    .sendJson({ type: 'connection_init', payload: {} })
    .expectJson({ type: 'connection_ack' }) // I assume this should go up here rather than after the start command
    .sendJson({ id: '1', type: 'start', payload: { operationName: 'catUpdated', query: /* ... */ } })
    .exec(() => request(app.getHttpServer())
      .post('/graphql')
      .send({
        operationName: 'updateCat',
        query: /* ... */,
        variables: { id, data: { name } },
      })
      .expect(200)
      .expect(({ body }) => { /* ... */ })
    )
    .expectJson((message) => {
      expect(message.type).toEqual('data');
      expect(message.payload.data).toEqual({ catUpdated: { id, name } });
    })
    .close();
});
```

To make things a bit clearer, you could of course extract the function passed into `exec`:

```javascript
const performUpdate = () => request(app.getHttpServer())
  .post('/graphql')
  .send({ /* ... */ })
  .expect(200)
  .expect(({ body }) => { /* ... */ });

// ...

await request(app.getHttpServer())
  .ws('/graphql', ['graphql-ws'])
  .sendJson({ type: 'connection_init', payload: {} })
  // ...
  .exec(performUpdate)
  .expectJson(/* ... */)
  .close();
```

**Comment by Shinigami92** [comment]
![Screenshot from 2022-02-09 09-15-43](https://user-images.githubusercontent.com/7195563/153154025-3953a354-4746-4623-9aec-3c5b3bd6b7a9.png)

Getting this error with your way :thinking: 

```node
 FAIL  test/cats.e2e-spec.ts
  ● Test suite failed to run

    test/cats.e2e-spec.ts:164:9 - error TS2322: Type 'Test' is not assignable to type 'void | Promise<void>'.
      Type 'Test' is not assignable to type 'Promise<void>'.
        Types of property 'then' are incompatible.
          Type '<TResult1 = Response, TResult2 = never>(onfulfilled?: ((value: Response) => TResult1 | PromiseLike<TResult1>) | null | undefined, onrejected?: ((reason: any) => TResult2 | PromiseLike<...>) | null | undefined) => Promise<...>' is not assignable to type '<TResult1 = void, TResult2 = never>(onfulfilled?: ((value: void) => TResult1 | PromiseLike<TResult1>) | null | undefined, onrejected?: ((reason: any) => TResult2 | PromiseLike<...>) | null | undefined) => Promise<...>'.
            Types of parameters 'onfulfilled' and 'onfulfilled' are incompatible.
              Types of parameters 'value' and 'value' are incompatible.
                Type 'Response' is not assignable to type 'void'.

    164         request(app.getHttpServer())
                ~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    165           .post('/graphql')
        ~~~~~~~~~~~~~~~~~~~~~~~~~~~
    ... 
    189             });
        ~~~~~~~~~~~~~~~
    190           }),
        ~~~~~~~~~~~~

      node_modules/.pnpm/superwstest@2.0.0_supertest@6.2.2/node_modules/superwstest/index.d.ts:45:14
        45     exec(fn: (ws: WebSocket) => Promise<void> | void): this;
                        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
        The expected type comes from the return type of this signature.
```

**Comment by davidje13** [comment]
@Shinigami92 seems to be unable to cast the promise return type to void, presumably because supertest's `Test` is promise-like, not an actual promise (probably a bug in the [TypeScript definitions for supertest](https://www.npmjs.com/package/@types/supertest)). You can work around it by adding explicit `async` / `await` syntax:

```javascript
    .exec(async () => {
      await request(app.getHttpServer())
        .post('/graphql')
        // ...
        .expect(200)
        .expect(({ body }) => { /* ... */ });
    })
```

**Comment by Shinigami92** [comment]
Will try that out tomorrow and report if it worked

**Comment by Shinigami92** [comment]
Mh the current solution I use works for now, and I also do some cleanup after the process.
So if I would return the whole block, I would need to wrap everything in nested functions and that would make the code a bit more complex.
So I will stick for my solution for now and hope that it will work in the future. if not, I can try to change or workaround it anyways.

**Comment by davidje13** [comment]
Turns out the `Promise<void>` issue is actually a bug in TypeScript itself (https://github.com/microsoft/TypeScript/issues/12871)

I pushed version 2.0.1 which switches to using `Promise<unknown>` to work around the issue. I also added some guidance to the bottom of the readme about this use-case.

**Comment by Shinigami92** [comment]
Now I have 2 websocket subscriptions :eyes:
How would I need to test this?

<details>
  <summary>Here is a full test example:</summary>

```ts
import { faker } from '@faker-js/faker';
import { HttpService } from '@nestjs/axios';
import type { INestApplication } from '@nestjs/common';
import { ValidationPipe } from '@nestjs/common';
import type { TestingModule } from '@nestjs/testing';
import { Test } from '@nestjs/testing';
import { of } from 'rxjs';
import { AppModule } from 'src/app.module';
import type { Cat } from 'src/cats/models/cat.model';
import { UUID_V4_REGEX } from 'src/common/utils/uuid.util';
import { createRandomCat } from 'src/mocks/cats';
import request from 'superwstest';

describe('CatsResolver (e2e)', () => {
  let app: INestApplication;

  const catsBaseUrl = 'https://cats.com';
  let httpService: HttpService;

  beforeEach(async () => {
    const moduleFixture: TestingModule = await Test.createTestingModule({
      imports: [AppModule],
    }).compile();

    app = moduleFixture.createNestApplication();
    app.useGlobalPipes(new ValidationPipe());
    await app.init();

    await app.listen(0, '0.0.0.0');

    httpService = app.get(HttpService);
  });

  afterEach(async () => {
    await app.close();
  });

  it('addCat', async () => {
    const name = 'New Cat';
    let createdCat: Cat;

    const httpSpy = jest
      .spyOn(httpService, 'post')
      .mockImplementation((path, data) => {
        expect(path).toEqual(catsBaseUrl + '/api/');
        expect(data).toEqual(
          expect.objectContaining({ name }),
        );
        createdCat = {
          id: faker.datatype.uuid(),
          name,
          createdAt: new Date(),
        };
        return of({
          data: createdCat,
          status: 200,
          statusText: 'OK',
          headers: {},
          config: {},
        });
      });

    const subscriptionTest = request(app.getHttpServer())
      .ws('/graphql', ['graphql-ws'])
      .sendJson({ type: 'connection_init', payload: {} })
      .sendJson({
        id: '1',
        type: 'start',
        payload: {
          operationName: 'catAdded',
          query: `#graphql
          subscription catAdded {
            catAdded {
              id
              name
              createdAt
            }
          }`,
        },
      })
      .expectJson({ type: 'connection_ack' });

    await subscriptionTest;

    let id = '';

    await request(app.getHttpServer())
      .post('/graphql')
      .send({
        operationName: 'addCat',
        query: `#graphql
        mutation addCat($data: NewCatInput!) {
          addCat(newCatData: $data) {
            id
            name
            createdAt
          }
        }`,
        variables: {
          data: {
            name,
          },
        },
      })
      .expect(200)
      .expect(({ body }) => {
        const data = body.data;
        expect(data.addCat).toBeTruthy();
        expect(data.addCat).toEqual({
          id: expect.stringMatching(UUID_V4_REGEX),
          name,
          createdAt: expect.any(String),
        });
        id = data.addCat.id;
      });

    await subscriptionTest
      .expectJson((message) => {
        expect(message.type).toEqual('data');
        const data = message.payload.data;
        expect(data).toEqual({
          catAdded: {
            id,
            name,
            createdAt: expect.any(String),
          },
        });
      })
      .close()
      .expectClosed();

    httpSpy.mockClear();
  });

  it('cats', () => {
    const totalCount = 50;

    const httpSpy = jest
      .spyOn(httpService, 'get')
      .mockImplementation((path) => {
        expect(path).toMatch(catsBaseUrl + '/api/');
        return of({
          data: path.endsWith('/count')
            ? totalCount
            : Array.from({ length: totalCount }, () => createRandomCat()),
          status: 200,
          statusText: 'OK',
          headers: {},
          config: {},
        });
      });

    const test = request(app.getHttpServer())
      .post('/graphql')
      .send({
        operationName: 'cats',
        query: `#graphql
        query cats {
          cats {
            nodes {
              id
              name
            }
            pageInfo {
              totalCount
            }
          }
        }`,
      })
      .expect(200)
      .expect(({ body }) => {
        const data = body.data;
        expect(data.cats.nodes[0].name).toEqual(expect.any(String));
        expect(data.cats.pageInfo.totalCount).toEqual(totalCount);
      });

    httpSpy.mockClear();

    return test;
  });

  it('updateCat', async () => {
    let cat: Cat;
    let id = '';

    const httpGetSpy = jest
      .spyOn(httpService, 'get')
      .mockImplementation((path) => {
        expect(path).toEqual(`${catsBaseUrl}/api/`);
        cat = createRandomCat();
        id = cat.id;
        return of({
          data: cat,
          status: 200,
          statusText: 'OK',
          headers: {},
          config: {},
        });
      });

    const name = 'Updated Cat';

    const httpUpdateSpy = jest
      .spyOn(httpService, 'put')
      .mockImplementation((path, data) => {
        expect(path).toEqual(`${catsBaseUrl}/api/`);
        expect(data).toEqual({ name });
        cat.name = data.name;
        return of({
          data: cat,
          status: 200,
          statusText: 'OK',
          headers: {},
          config: {},
        });
      });

    const subscriptionTest = request(app.getHttpServer())
      .ws('/graphql', ['graphql-ws'])
      .sendJson({ type: 'connection_init', payload: {} })
      .sendJson({
        id: '1',
        type: 'start',
        payload: {
          operationName: 'catUpdated',
          query: `#graphql
          subscription catUpdated {
            catUpdated {
              id
              name
            }
          }`,
        },
      })
      .expectJson({ type: 'connection_ack' });

    await subscriptionTest;

    await request(app.getHttpServer())
      .post('/graphql')
      .send({
        operationName: 'updateCat',
        query: `#graphql
        mutation updateCat($id: String!, $data: UpdateCatInput!) {
          updateCat(id: $id, updateCatData: $data) {
            id
            name
          }
        }`,
        variables: {
          id,
          data: {
            name,
          },
        },
      })
      .expect(200)
      .expect(({ body }) => {
        const data = body.data;
        expect(data.updateCat).toBeTruthy();
        expect(data.updateCat).toEqual({
          id,
          name,
        });
      });

    await subscriptionTest
      .expectJson((message) => {
        expect(message.type).toEqual('data');
        const data = message.payload.data;
        expect(data).toEqual({
          catUpdated: {
            id,
            name,
          },
        });
      })
      .close()
      .expectClosed();

    httpGetSpy.mockClear();
    httpUpdateSpy.mockClear();
  });

  it('removeCat', async () => {
    const id = faker.datatype.uuid();

    const httpSpy = jest
      .spyOn(httpService, 'delete')
      .mockImplementationOnce((path) => {
        expect(path).toEqual(`${catsBaseUrl}/api/${id}`);
        return of({
          data: true,
          status: 200,
          statusText: 'OK',
          headers: {},
          config: {},
        });
      });

    const subscriptionTest = request(app.getHttpServer())
      .ws('/graphql', ['graphql-ws'])
      .sendJson({ type: 'connection_init', payload: {} })
      .sendJson({
        id: '1',
        type: 'start',
        payload: {
          operationName: 'catRemoved',
          query: `#graphql
          subscription catRemoved {
            catRemoved
          }`,
        },
      })
      .expectJson({ type: 'connection_ack' });

    await subscriptionTest;

    await request(app.getHttpServer())
      .post('/graphql')
      .send({
        operationName: 'removeCat',
        query: `#graphql
        mutation removeCat($id: String!) {
          removeCat(id: $id)
        }`,
        variables: {
          id,
        },
      })
      .expect(200)
      .expect(({ body }) => {
        const data = body.data;
        expect(data.removeCat).toBeTruthy();
      });

    await subscriptionTest
      .expectJson((message) => {
        expect(message.type).toEqual('data');
        const data = message.payload.data;
        expect(data).toEqual({
          catRemoved: id,
        });
      })
      .close()
      .expectClosed();

    httpSpy.mockClear();
  });
});
```
</details>

**Comment by davidje13** [comment]
@Shinigami92 I actually added an example for exactly that at the very bottom of the readme when I released 2.0.1: https://github.com/davidje13/superwstest#how-can-i-perform-another-asynchronous-task-while-a-connection-is-open (3rd example)

The gist is that you use `Promise.all` to run multiple chains simultaneously. You can set it up so that one connection will be waiting for the effect of an action (via an `expect`), while the other is performing the action (via `send`).

If you need more complicated synchronisation between the two connections (more complicated than each one waiting for an action caused by the other) you'll need to add your own synchronisation logic (you can use `.exec` as the hook for this but the actual syncing will be up to you; I'd recommend something like an asynchronous blocking queue acting as a semaphore — the details are beyond the scope of this library though)
